core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1652,25 +1652,28 @@ private[spark] object Utils extends Logging {
    * Attempt to start a service on the given port, or fail after a number of attempts.
    * Each subsequent attempt uses 1 + the port used in the previous attempt (unless the port is 0).
    *
-   * @param startPort The initial port to start the service on.
+   * @param port The initial port to start the service on.
    * @param maxRetries Maximum number of retries to attempt.
    *                   A value of 3 means attempting ports n, n+1, n+2, and n+3, for example.
    * @param startService Function to start service on a given port.
    *                     This is expected to throw java.net.BindException on port collision.
    */
   def startServiceOnPort[T](
-      startPort: Int,
+      port: Int,
       startService: Int => (T, Int),
+      conf: SparkConf = new SparkConf(),
       serviceName: String = "",
       maxRetries: Int = portMaxRetries): (T, Int) = {
     val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
+    val startPort = conf.getInt("spark.port.min", 1024)
+    val endPort = conf.getInt("spark.port.max", 65536)
     for (offset <- 0 to maxRetries) {
-      // Do not increment port if startPort is 0, which is treated as a special port
-      val tryPort = if (startPort == 0) {
-        startPort
+      // Do not increment port if port is 0, which is treated as a special port
+      val tryPort = if (port == 0) {
+        port
```

Comment:
sorry I think your last comment/question hits on this issue.  that seems better.  As long as all the services default to port 0 (other then web ui) this seems fine.  That way if user does specify a port explicitly it will still use that.